### PR TITLE
Allow custom mono modules to search for, and a custom name for the _Data folder

### PR DESCRIPTION
**Pull request by mitchell-merry**
# Description
We need to additional configuration around the `_Data` folder and the mono module names.

This PR adds that.

# Testing
Got users of the autosplitter to test [this version](https://github.com/LiterallyMetaphorical/LiveSplit.POPTLC/blob/05a5ee48cac1eaa776d2472c8010937a3f26d365/POPTLC.asl#L6-L10).

```cs
startup
{
    Assembly.Load(File.ReadAllBytes("Components/asl-help")).CreateInstance("Unity");
    vars.Helper.IL2CPPModules = new List<string>() { "GameAssembly.dll", "GameAssembly_plus.dll" };
    vars.Helper.DataFolder = "TheLostCrown_Data";
}
```

This allows the autosplitter to find the mono module regardless of if it is GameAssembly.dll or GameAssembly_plus.dll, which it is for the Ubisoft+ version of the game. Specifying an override for DataFolder is also crucial for some points in the code, so that was done too.

# Risk
No. Logic remains the same without changing these.

---
- [x] I have read and accept asl-help's license.

### CODE REVIEW

**Comment by ero-qt** [comment]
maybe just a string for `MonoModule` would be enough. can set it in startup or init. if unset (null), infer automatically. if set, assume the user knows what they're doing. if they mess up, their fault.

**Comment by mitchell-merry** [comment]
Thought you didn't care :P

I need (well, want) to specify multiple because different versions have different mono modules and I don't want to dynamically choose

There's definitely better ways to implement this, it was just quick and dirty

File: src/Basic/Validation.cs [hunk 1]
```diff
@@ -0,0 +1,24 @@
+using AslHelp.Data.AutoSplitter;
+
+namespace AslHelp;
+
+public static class Validation
```
Comment (ero-qt) [inline src/Basic/Validation.cs]: call this `Assert`

File: src/Basic/Validation.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+{
+    public static void AssertNotNull(string name, dynamic value)
```
Comment (ero-qt) [inline src/Basic/Validation.cs]: call this `NotNull`. use `object` instead of `dynamic`. use `CallerArgumentExpressionAttribute` for automatic argument name inference.

File: src/Basic/Validation.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    {
+        if (value == null)
+        {
+            throw new ArgumentNullException($"{name} may not be null.");
+        }
+    }
+
+    public static void AssertAction(string name, string action)
```
Comment (ero-qt) [inline src/Basic/Validation.cs]: call this `InAction`. use `CallerMemberNameAttribute` for automatic caller name inference.

File: src/Basic/Validation.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    {
+
+        if (Actions.Current != "startup")
```
Comment (ero-qt) [inline src/Basic/Validation.cs]:
```suggestion
        if (Actions.Current != action)
```

File: src/Unity/Helper/Unity.Data.cs [hunk 5]
```diff
@@ -10,23 +13,80 @@ public partial class Unity
...
-                throw new InvalidOperationException(msg);
-            }
+            Validation.AssertAction(nameof(LoadSceneManager), "startup");
 
             Debug.Info(value ? "  => Will try to load SceneManager." : "  => Will not load SceneManager.");
             _loadSceneManager = value;
         }
     }
 
+    private List<string> _monoV1Modules = ["mono.dll"];
```
Comment (ero-qt) [inline src/Unity/Helper/Unity.Data.cs]: why not an array?

File: src/Unity/Helper/Unity.Data.cs [hunk 6]
```diff
@@ -10,23 +13,80 @@ public partial class Unity
...
+    public List<string> MonoV1Modules
+    {
+        get => _monoV1Modules;
+        set
+        {
+            Validation.AssertAction(nameof(MonoV1Modules), "startup");
+            Validation.AssertNotNull(nameof(MonoV1Modules), value);
+
+            Debug.Info($"  => Will use the Mono V1 modules {string.Join(", ", value)}. ");
```
Comment (ero-qt) [inline src/Unity/Helper/Unity.Data.cs]: this isn't really necessary. the user sets these manually, they already know what's going to be used.  
yes, i have prints like this elsewhere, and they should also be reduced.

File: src/Unity/Helper/Unity.Data.cs [hunk 7]
```diff
@@ -10,23 +13,80 @@ public partial class Unity
     public Module UnityPlayer { get; private set; }
     public SceneManager Scenes { get; private set; }
 
+    private string _dataFolder;
+    public string DataFolder
```
Comment (ero-qt) [inline src/Unity/Helper/Unity.Data.cs]: please explain why this is necessary (ideally with examples).

File: src/Unity/Helper/Unity.Load.cs [hunk 8]
```diff
@@ -23,13 +19,13 @@ protected override UnityMemManager MakeManager()
 
         return MonoModule.Name switch
         {
-            MONO_V1 => new MonoV1Manager("v1"),
-            MONO_V2 => (UnityVersion.Major, UnityVersion.Minor) switch
+            var _ when MonoV1Modules.Contains(MainModule.Name) => new MonoV1Manager("v1"),
```
Comment (ero-qt) [inline src/Unity/Helper/Unity.Load.cs]: the switch is pointless if all cases are just `Contains` checks.

File: src/Basic/Validation.cs (hunk 2 above)
Comment (mitchell-merry) [reply src/Basic/Validation.cs]: https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.callerargumentexpressionattribute?view=net-8.0 appears to be in a future version of .net, am I wrong on this?

File: src/Unity/Helper/Unity.Data.cs (hunk 7 above)
Comment (mitchell-merry) [reply src/Unity/Helper/Unity.Data.cs]: For Prince of Persia: The Lost Crown's default version, the .exe is `TheLostCrown.exe`, the mono module is `GameAssembly.dll`, and the data folder is `TheLostCrown_Data`.

For the Ubisoft+ version of the game, the .exe is `TheLostCrown_plus.exe`, the mono module is `GameAssembly_plus.dll`, and the data folder is still `TheLostCrown_Data`.

The code currently infers the name of the data folder from the main module (`TheLostCrown_plus.exe`), so it looks for a folder which does not exist (`TheLostCrown_plus_Data`). This lets us override the folder it looks for in case it is different.

File: src/Basic/Validation.cs (hunk 2 above)
Comment (ero-qt) [reply src/Basic/Validation.cs]: annoying. polysharp fixes this, but introducing that to asl-help would open up way too many things that i won't deal with (nullability). i already use it extensively in v2, but won't bother with it here.

Comment (ero-qt) [inline src/Basic/Validation.cs]:
```suggestion
    public static void NotNull(object argument, string paramName)
```

**Comment by mitchell-merry** [comment]
currently testing changes

File: src/Basic/Validation.cs [hunk 9]
```diff
@@ -1,21 +1,22 @@
...
+public static class Assert
 {
-    public static void AssertNotNull(string name, dynamic value)
+    public static void NotNull(object argument, string paramName)
     {
-        if (value == null)
+        if (argument == null)
         {
-            throw new ArgumentNullException($"{name} may not be null.");
+            throw new ArgumentNullException($"{argument} may not be null.");
```
Comment (ero-qt) [inline src/Basic/Validation.cs]:
```suggestion
            throw new ArgumentNullException($"{paramName} may not be null.");
```

File: src/Basic/Validation.cs [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+    public static void InAction(string action, [CallerMemberName] string name = "")
+    {
+
+        if (Actions.Current != action)
+        {
+            string msg = $"{name} may only be set in the '{action} {{}}' action.";
+            throw new InvalidOperationException(msg);
+        }
+    }
```
Comment (ero-qt) [inline src/Basic/Validation.cs]:
```suggestion
    public static void InAction(string action, [CallerMemberName] string callerName = "")
    {

        if (Actions.Current != action)
        {
            string msg = $"{callerName} may only be set in the '{action} {{}}' action.";
            throw new InvalidOperationException(msg);
        }
    }
```

**Comment by mitchell-merry** [comment]
is working, at least the happy case is

File: src/Unity/Helper/Unity.Data.cs (hunk 7 above)
Comment (mitchell-merry) [reply src/Unity/Helper/Unity.Data.cs]: image

File: src/Unity/Helper/Unity.Data.cs [hunk 11]
```diff
@@ -10,23 +13,77 @@ public partial class Unity
...
+        set
+        {
+            Assert.InAction("startup");
+            Assert.NotNull(value, nameof(MonoV1Modules));
+
+            _monoV1Modules = value;
+        }
+    }
+
+    private string[] _monoV2Modules = ["GameAssembly.dll"];
```
Comment (mitchell-merry) [inline src/Unity/Helper/Unity.Data.cs]: typo dont merge

**Comment by ero-qt** [comment]
Considering we don't know all of the edge cases, I would prefer a much simpler implementation. We don't know whether the `_Data` directory will always be the same for other games which may take this approach.

It would be much easier to simply set the Mono module name and the `_Data` directory name on a case-by-case basis.

Code such as this;
```cs
init
{
  if (game.ProcessName == "TheLostCrown_plus")
  {
    vars.Helper.MonoModuleName = "GameAssembly_plus.dll";
    vars.Helper.DataDirectory = "TheLostCrown_Data";
  }

  // ...
}
```
hardly increases complexity and achieves the same thing while being a lot more deliberate.

**Comment by ero-qt** [comment]
Please test with other games and build with `-c Release`.

**Comment by mitchell-merry** [comment]
> Considering we don't know all of the edge cases, I would prefer a much simpler implementation. We don't know whether the `_Data` directory will always be the same for other games which may take this approach.
> 
> It would be much easier to simply set the Mono module name and the `_Data` directory name on a case-by-case basis.
> 
> Code such as this;
> 
> ```cs
> init
> {
>   if (game.ProcessName == "TheLostCrown_plus")
>   {
>     vars.Helper.MonoModuleName = "GameAssembly_plus.dll";
>     vars.Helper.DataDirectory = "TheLostCrown_Data";
>   }
> 
>   // ...
> }
> ```
> 
> hardly increases complexity and achieves the same thing while being a lot more deliberate.

Given a `vars.Helper.MonoModuleName = "GameAssembly_plus.dll";`, how do you determine what type of manager to initialise? (IL2CPP, mono v1, etc)

**Comment by ero-qt** [comment]
> Given a `vars.Helper.MonoModuleName = "GameAssembly_plus.dll";`, how do you determine what type of manager to initialise? (IL2CPP, mono v1, etc)

This is a good question to think about, even for v2. The first thing that comes to mind is checking whether `GameName_Data/il2cpp_data/metadata/global-metadata.dat` is present.

**Comment by mitchell-merry** [comment]
And mono v1/v2? I'm not sure that looking for that file is great, although it's workable. 

Unless there's a good implementation for this right now, I'd prefer to move the PR along. This change is blocking some people for the sr.c challenge, and it's likely only this autosplitter will actually use the feature.

Much appreciate your quick responses and detailed review

File: src/Unity/Helper/Unity.Data.cs [hunk 12]
```diff
@@ -10,23 +13,77 @@ public partial class Unity
...
+            string directory = Path.GetDirectoryName(MainModule.FilePath);
+            string folderName = _dataFolder ?? Path.GetFileName(MainModule.FilePath)[..^4];
+
+            return Path.Combine(directory, folderName);
+        }
+        set
+        {
+            Assert.InAction("startup");
+
+            Debug.Info($"  => Will use {value} as the DataFolder.");
```
Comment (ero-qt) [inline src/Unity/Helper/Unity.Data.cs]: please remove this as well

File: src/Unity/Helper/Unity.Data.cs [hunk 13]
```diff
@@ -10,23 +13,77 @@ public partial class Unity
...
 
+    private string _dataFolder;
+    public string DataFolder
+    {
+        get
+        {
+            string directory = Path.GetDirectoryName(MainModule.FilePath);
+            string folderName = _dataFolder ?? Path.GetFileName(MainModule.FilePath)[..^4];
+
+            return Path.Combine(directory, folderName);
```
Comment (ero-qt) [inline src/Unity/Helper/Unity.Data.cs:21]: please check whether this output is correct on other games. it does not appear to me that this will result in the output you expect.

File: src/Unity/Helper/Unity.Data.cs [hunk 14]
```diff
@@ -10,23 +13,77 @@ public partial class Unity
     public Module UnityPlayer { get; private set; }
     public SceneManager Scenes { get; private set; }
 
+    private string _dataFolder;
+    public string DataFolder
```
Comment (ero-qt) [inline src/Unity/Helper/Unity.Data.cs]:
```suggestion
    private string _dataDirectory;
    public string DataDirectory
```

**Comment by ero-qt** [comment]
Please understand that I simply want to clean this up to conform to the existing codebase (and my personal coding standards).

File: src/Unity/Helper/Unity.Data.cs [hunk 15]
```diff
@@ -10,23 +13,77 @@ public partial class Unity
...
-                throw new InvalidOperationException(msg);
-            }
+            Assert.InAction("startup");
 
             Debug.Info(value ? "  => Will try to load SceneManager." : "  => Will not load SceneManager.");
             _loadSceneManager = value;
         }
     }
 
+    private string[] _monoV1Modules = ["mono.dll"];
```
Comment (ero-qt) [inline src/Unity/Helper/Unity.Data.cs]: potential idea here; turn this back into a `List<string>`, make the property `get`-only, and have the user (you) do `MonoV1Modules.Add("")`.

**Comment by ero-qt** [comment]
This should be all. After these are resolved, please build with `-c Release` and push again.

**Comment by mitchell-merry** [comment]
> Please understand that I simply want to clean this up to conform to the existing codebase (and my personal coding standards).

It's no problem!

> This should be all. After these are resolved, please build with -c Release and push again.

After your changes, it is still working, with and without the config as expected (I'm running a version of the game that doesn't need the config), and it seems from reports that the config does enable it to work for others.

So let's get this merged?

File: src/Unity/Helper/Unity.Data.cs (hunk 13 above)
Comment (mitchell-merry) [reply src/Unity/Helper/Unity.Data.cs:21]: It works for me even if I don't specify it, so you'll have to be more specific with what you mean.

Comment (mitchell-merry) [reply src/Unity/Helper/Unity.Data.cs:21]: I have tried on other games and it all works
